unix/x0vncserver/XDesktop.cxx

```diff
@@ -383,6 +391,92 @@ KeyCode XDesktop::XkbKeysymToKeycode(Display* dpy, KeySym keysym) {
 }
 #endif
 
+KeyCode XDesktop::addKeysym(Display* dpy, KeySym keysym)
+{
+  unsigned int key;
+  XkbDescPtr xkb;
+  XkbMapChangesRec changes;
+  KeySym *syms;
+  xkb = XkbGetMap(dpy, XkbAllComponentsMask, XkbUseCoreKbd);
+
+  if (!xkb)
+    return false;
+
+  for (key = xkb->max_key_code; key >= xkb->min_key_code; key--) {
+    if (XkbKeyNumGroups(xkb, key) == 0)
+      break;
+  }
+
+  if (key < xkb->min_key_code)
+    return false;
+
+  memset(&changes, 0, sizeof(changes));
+
+  syms = XkbKeySymsPtr(xkb, key);
+  syms[0] = keysym;
+
+  int types[XkbNumKbdGroups] = { XkbOneLevelIndex };
+  XkbChangeTypesOfKey(xkb, key, 1, XkbGroup1Mask, types, &changes);
+  XkbKeySymEntry(xkb, key, 0, 0) = keysym;
+
+  changes.changed |= XkbKeySymsMask;
+  changes.first_key_sym = key;
+  changes.num_key_syms = 1;
+
+  if (XkbChangeMap(dpy, xkb, &changes)) {
+    vlog.info("Added unknown keysym 0x%lx to keycode %d", keysym, key);
+    added_keysyms[key] = keysym;
+    return key;
+  }
+
+  return 0;
+}
+
+void XDesktop::deleteAddedKeycodes(Display* dpy) {
+  XkbDescPtr xkb;
+  xkb = XkbGetMap(dpy, XkbAllComponentsMask, XkbUseCoreKbd);
+
+  if (!xkb)
+    return;
+
+  for (int key = xkb->min_key_code; key < xkb->max_key_code; key++) {
+    if (added_keysyms[key] != NoSymbol && XkbKeyNumGroups(xkb, key) != 0) {
```

Comment:
This will wipe any changes the user might have made. We should probably try to confirm that it is our changes that we are deleting.

Reply:
Shouldn't all added keysyms in this `vector` (now `map`) be ours? I always put it there only when we successfuly add a new keysym.

Reply:
No, because other tools might also update the keyboard layout. E.g. `setxkbmap`. x0vncserver doesn't have exclusive control of the layout.